Many tests under imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/ are timing out

cdumez opened the pull request:
#### 71551c7e879e08324f10ea8a8f994ac64012b430
<pre>
Many tests under imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/ are timing out
<a href="https://bugs.webkit.org/show_bug.cgi?id=243503">https://bugs.webkit.org/show_bug.cgi?id=243503</a>

Reviewed by Darin Adler.

I investigated form-requestsubmit-during-load.html and what the test does is:
1. Create a &quot;sentinel&quot; iframe, Navigate it to URL1 and then again to URL2
2. Create a test iframe to a page which submits a form to navigate the frame in the load event handler
3. Call `history.back()` in the main frame

Because the test iframe initiated a navigation (via form submission) from its load event handler,
we are expected NOT to create a HistoryItem for this navigation and instead replace the current
one (which WebKit gets right).

At step 3, because the test frame didn&apos;t actually push a new HistoryItem, the navigation back
should navigate the sentinel iframe from URL2 back to URL1. When attempting this navigation,
WebKit would instead reload the whole page, causing the test to loop infinitely.

To give some background, whenever we navigate a subframe on the page, WebKit clones the
top frame&apos;s HistoryItem. Each top-level HistoryItem has child HistoryItems for each of
its subframes. So when navigating a subframe, you would expect a clone of the top-level
HistoryItem and then the URL of one of its child HistoryItems to get updated.

In WebKit, HistoryItems that are clones have the same itemSequenceNumber to make it clear
when we&apos;re navigating within the same page. When HistoryController:recursiveGoToItem()
gets called, it checks recursively the source and destination HistoryItems and is only
meant to navigate the frames that have actually changed URL between the two HistoryItems.
To detect if 2 HistoryItems are clones (and thus no navigation of this frame needs to
occur), we rely on HistoryController::itemsAreClones(). itemsAreClones() would check
that the pointers are different but that the itemSequenceNumber actually match, as is
expected. However, itemsAreClones() was doing some extra checks that apparently didn&apos;t
match other browser and were causing us to fail those tests. In particular, it was
checking that the two clone items have the exact same iframes. In the context of the
test, one of the HistoryItems has one iframe and the other had two, causing the
check to fail and we would thus reload the top HistoryItem / frame. This is likely
caused by the fact that the second iframe did a HistoryItem replacement because its
first load completed.

To align we other browsers and pass the WPT tests (which all pass in Blink, as most of
them do in Gecko too), I have dropped the extra checks and only rely on the pointer
value and the itemSequenceNumber. When trying to check if a particular subframe needs
to be navigated, we check then that the iframe in question is present in both the
source and destination HistoryItem (instead of asserting that this is the case).

* LayoutTests/TestExpectations:
* LayoutTests/fast/history/history-subframe-with-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-requestsubmit-during-load-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-requestsubmit-during-pageshow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-requestsubmit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-submit-button-click-during-load-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-submit-button-click-during-pageshow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-submit-button-click-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-submit-during-load-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-submit-during-pageshow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-submit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/iframe-src-during-load-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/iframe-src-during-pageshow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/iframe-src-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/location-assign-during-load-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/location-assign-during-pageshow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/location-assign-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/location-setter-during-load-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/location-setter-during-pageshow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/location-setter-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::FrameLoader::HistoryController::recursiveSetProvisionalItem):
(WebCore::FrameLoader::HistoryController::recursiveGoToItem):
(WebCore::FrameLoader::HistoryController::itemsAreClones const):

Canonical link: <a href="https://commits.webkit.org/253099@main">https://commits.webkit.org/253099@main</a>
</pre>

darinadler reviewed (approved):
Sounds good to me; seems like there is some compatibility risk that is mitigated by the WPT tests and Chrome’s behavior.

cdumez committed:
Many tests under imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/ are timing out
https://bugs.webkit.org/show_bug.cgi?id=243503

Reviewed by Darin Adler.

I investigated form-requestsubmit-during-load.html and what the test does is:
1. Create a "sentinel" iframe, Navigate it to URL1 and then again to URL2
2. Create a test iframe to a page which submits a form to navigate the frame in the load event handler
3. Call `history.back()` in the main frame

Because the test iframe initiated a navigation (via form submission) from its load event handler,
we are expected NOT to create a HistoryItem for this navigation and instead replace the current
one (which WebKit gets right).

At step 3, because the test frame didn't actually push a new HistoryItem, the navigation back
should navigate the sentinel iframe from URL2 back to URL1. When attempting this navigation,
WebKit would instead reload the whole page, causing the test to loop infinitely.

To give some background, whenever we navigate a subframe on the page, WebKit clones the
top frame's HistoryItem. Each top-level HistoryItem has child HistoryItems for each of
its subframes. So when navigating a subframe, you would expect a clone of the top-level
HistoryItem and then the URL of one of its child HistoryItems to get updated.

In WebKit, HistoryItems that are clones have the same itemSequenceNumber to make it clear
when we're navigating within the same page. When HistoryController:recursiveGoToItem()
gets called, it checks recursively the source and destination HistoryItems and is only
meant to navigate the frames that have actually changed URL between the two HistoryItems.
To detect if 2 HistoryItems are clones (and thus no navigation of this frame needs to
occur), we rely on HistoryController::itemsAreClones(). itemsAreClones() would check
that the pointers are different but that the itemSequenceNumber actually match, as is
expected. However, itemsAreClones() was doing some extra checks that apparently didn't
match other browser and were causing us to fail those tests. In particular, it was
checking that the two clone items have the exact same iframes. In the context of the
test, one of the HistoryItems has one iframe and the other had two, causing the
check to fail and we would thus reload the top HistoryItem / frame. This is likely
caused by the fact that the second iframe did a HistoryItem replacement because its
first load completed.

To align we other browsers and pass the WPT tests (which all pass in Blink, as most of
them do in Gecko too), I have dropped the extra checks and only rely on the pointer
value and the itemSequenceNumber. When trying to check if a particular subframe needs
to be navigated, we check then that the iframe in question is present in both the
source and destination HistoryItem (instead of asserting that this is the case).

* LayoutTests/TestExpectations:
* LayoutTests/fast/history/history-subframe-with-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-requestsubmit-during-load-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-requestsubmit-during-pageshow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-requestsubmit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-submit-button-click-during-load-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-submit-button-click-during-pageshow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-submit-button-click-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-submit-during-load-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-submit-during-pageshow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/form-submit-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/iframe-src-during-load-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/iframe-src-during-pageshow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/iframe-src-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/location-assign-during-load-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/location-assign-during-pageshow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/location-assign-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/location-setter-during-load-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/location-setter-during-pageshow-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/replace-before-load/location-setter-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::FrameLoader::HistoryController::recursiveSetProvisionalItem):
(WebCore::FrameLoader::HistoryController::recursiveGoToItem):
(WebCore::FrameLoader::HistoryController::itemsAreClones const):

Canonical link: https://commits.webkit.org/253099@main
